doc/ml_features.rst

```diff
@@ -0,0 +1,337 @@
+result can be achieved with any SmartRedis client (C, C++,
+Fortran and Python). Please refer to SmartRedis API
+for language-specific details.
+
+Using ML models on the DB
+=========================
+
+The combination of SmartSim and SmartRedis allows users
+to store more than simple tensors on the DB. In the following
+subsections, we show how to upload executable code, in the
+form of ML models or functions, to the DB.
+The stored code can then be run on stored tensors, and
+the output is stored on the DB as well, where it can be
+downloaded with standard ``get_tensor`` calls.
+
+In general, there are two ways to upload serialized code
+to the DB: from memory and from file. In all examples, we
+will assume that a SmartSim ``Orchestator`` is up and running,
+and that the code we will show is run as part of a SmartSim-launched
+application ``Model``.
+
+
+TensorFlow
+----------
+SmartSim provides :ref:`two helper methods for serializing
+TensorFlow and Keras models <smartsim_tf_api>`: ``freeze_model`` and
+``serialize_model``.
+
+The method ``freeze_model`` is thought to be used in conjunction
+with SmartRedis ``set_model_from_file``. The following is a typical
+workflow, first we define the model:
+
+.. code-block:: python
+
+    import numpy as np
+    from smartredis import Client
+    from tensorflow import keras
+    from smartsim.ml.tf import freeze_model
+
+    model = keras.Sequential(
+        layers=[
+            keras.layers.InputLayer(input_shape=(28, 28), name="input"),
+            keras.layers.Flatten(input_shape=(28, 28), name="flatten"),
+            keras.layers.Dense(128, activation="relu", name="dense"),
+            keras.layers.Dense(10, activation="softmax", name="output"),
+        ],
+        name="FCN",
+    )
+
+    # Compile model with optimizer
+    model.compile(
+        optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+
+The model is now ready to be serialized and stored on the filesystem:
+
+.. code-block:: python
+
+    model_path, inputs, outputs = freeze_model(model, '.', "mnist.pb")
+
+Note that ``freeze_model`` conveniently returns the path to the serialized model file,
+and the names of the input and output layers, which are needed to upload the TensorFlow
+model on the DB, as shown in the following code snippet, where we also upload a
+synthetic sample to be passed to the model. Notice that we could also upload a batch
+of samples, instead of a single one.
+
+.. code-block:: python
+
+    client = Client(cluster=False)
+    model_key = "tf_mnist"
+    client.set_model_from_file(
+        model_key, model_path, "TF", device="GPU", inputs=inputs, outputs=outputs
+    )
+    mnist_image = np.random.rand(1, 28, 28).astype(np.float32)
+    client.put_tensor("mnist_input", mnist_image)
+
+For details about ``set_model_from_file``, please
+refer to :ref:`SmartRedis API <smartredis-api>`.
+Finally, the model can be run on the sample and the output is ready to be downloaded.
+
```

Comment:
I suggest breaking up the examples even more:

```
|- Tensorflow
   **add short summary here**
   |- serialize the model
   |- upload from file system
   |- upload from memory
   |- run the model
   |- source code
|- PyTorch
   **add short summary here**
   |- jittrace the model
   |- upload from file system
   |- upload from memory
   |- run the model
   |- source code
|- TorchScript Functions
   **add short summary here**
   |- prep model
   |- upload from file system
   |- upload from memory
   |- run the model
   |- source code
|- ONNX Runtime
   **add short summary here**
   |- prep model
   |- upload from file system
   |- upload from memory
   |- run the model
   |- source code
```

Feel free to ignore, I was thinking that we could make parts more distinguished tho in some way!